src/tactic/move_add.lean

```diff
@@ -0,0 +1,400 @@
+        return (summed, list_unused.transpose.map list.band)
+    | none := do
+        [(Fn, unused_F), (bn, unused_b)] ← [F, b].mmap $ reorder_oper,
+        return $ (expr.app Fn bn, [unused_F, unused_b].transpose.map list.band)
+    end
+| (expr.pi na bi e f)           := do
+  [en, fn] ← [e, f].mmap $ reorder_oper,
+  return (expr.pi  na bi en.1 fn.1, [en.2, fn.2].transpose.map list.band)
+| (expr.lam na bi e f)          := do
+  [en, fn] ← [e, f].mmap $ reorder_oper,
+  return (expr.lam na bi en.1 fn.1, [en.2, fn.2].transpose.map list.band)
+| (expr.mvar na pp e)           := do  -- is it really needed to recurse here?
+  en ← reorder_oper e,
+  return (expr.mvar na pp en.1, [en.2].transpose.map list.band)
+| (expr.local_const na pp bi e) := do  -- is it really needed to recurse here?
+  en ← reorder_oper e,
+  return (expr.local_const na pp bi en.1, [en.2].transpose.map list.band)
+| (expr.elet na e f g)          := do
+  [en, fn, gn] ← [e, f, g].mmap $ reorder_oper,
+  return (expr.elet na en.1 fn.1 gn.1, [en.2, fn.2, gn.2].transpose.map list.band)
+| (expr.macro ma le)            := do  -- is it really needed to recurse here?
+  len ← le.mmap $ reorder_oper,
+  let (lee, lb) := len.unzip,
+  return (expr.macro ma lee, lb.transpose.map list.band)
+| e := pure (e, (lp.map (λ _, tt)))
+
+/-- Passes the user input `na` to `reorder_oper` at a single location, that could either be
+`none` (referring to the goal) or `some name` (referring to hypothesis `name`).  Replaces the
+given hypothesis/goal with the rearranged one that `reorder_hyp` receives from `reorder_oper`.
+Returns a pair consisting of a boolean and a further list of booleans.
+The single boolean is `tt` iff the tactic did *not* change the goal on which it was acting.
+The list of booleans records which variable in `ll` has been unified in the application:
+`tt` means that the corresponding variable has *not* been unified.
+
+This definition is useful to streamline error catching. -/
+meta def reorder_hyp (op : pexpr) (lp : list (bool × pexpr)) (na : option name) :
+  tactic (bool × list bool) := do
+(thyp, hyploc) ← match na with
+  | none := do
+      t ← target,
+      return (t, none)
+  | some na := do
+      hl ← get_local na,
+      th ← infer_type hl,
+      return (th, some hl)
+  end,
+(reordered, is_unused) ← reorder_oper op lp thyp,
+unify reordered thyp >> return (tt, is_unused) <|> do
+-- the current `do` block takes place where the reordered expression is not equal to the original
+neq ← mk_app `eq [thyp, reordered],
+nop ← to_expr op tt ff,
+pre ← pp reordered,
+(_, prf) ← solve_aux neq $ match nop with
+  | `(has_add.add) := `[{ simp only [add_comm, add_assoc, add_left_comm]; refl, done }]
+  | `(has_mul.mul) := `[{ simp only [mul_comm, mul_assoc, mul_left_comm]; refl, done }]
+  | _ := ac_refl <|>
+    fail format!("the associative/commutative lemmas used do not suffice to prove that " ++
+      "the initial goal equals:\n\n{pre}\n")
```

Comment:
I think tactics should make clear when they fail whether the user can do something to fix the issue, or whether this is a bug in the tactic itself. How about this docstring?
```suggestion
    fail format!("the associative/commutative lemmas used do not suffice to prove that " ++
      "the initial goal equals:\n\n{pre}\n" ++
      "Hint: try adding `is_associative` or `is_commutative` instances.\n")
```